File: Engine/MonoScripting/Api/EntityComponents/ParticleEmitterComponent.cpp
#include "../ApiBindings.hpp"
#include <Scene/EntityComponents/ParticleEmitterComponent.hpp>
#include <MonoScripting/MonoScriptingLayer.hpp>

using namespace VSGE;

static void StartSimulation(ParticleEmitterComponent* pec) {
	pec->StartSimulation();
}

static void StopSimulation(ParticleEmitterComponent* pec) {
	pec->StopSimulation();
}

static bool IsSimulating(ParticleEmitterComponent* pec) {
	return pec->IsSimulating();
}

static float GetDuration(ParticleEmitterComponent* pec) {
	return pec->GetDuration();
}

static void SetDuration(ParticleEmitterComponent* pec, float duration) {
	pec->SetDuration(duration);
}

static bool IsLooping(ParticleEmitterComponent* pec) {
	return pec->GetLooping();
}

static void SetLooping(ParticleEmitterComponent* pec, bool loop) {
	pec->SetLooping(loop);
}

static bool IsPrewarm(ParticleEmitterComponent* pec) {
	return pec->GetPrewarm();
}

static void SetPrewarm(ParticleEmitterComponent* pec, bool prewarm) {
	pec->SetPrewarm(prewarm);
}

static float GetParticleLifetime(ParticleEmitterComponent* pec) {
	return pec->GetParticleLifetime();
}

static void SetParticleLifetime(ParticleEmitterComponent* pec, float lifetime) {
	pec->SetParticleLifetime(lifetime);
}

void VSGE::BindParticleEmitter() {
	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_StartSimulation(ulong)", StartSimulation);
	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_StopSimulation(ulong)", StopSimulation);
	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_IsSimulating(ulong)", IsSimulating);

	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_GetDuration(ulong)", GetDuration);
	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_SetDuration(ulong,float)", SetDuration);

	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_IsLooping(ulong)", IsLooping);
	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_SetLooping(ulong,bool)", SetLooping);

	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_IsPrewarm(ulong)", IsPrewarm);
	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_SetPrewarm(ulong,bool)", SetPrewarm);

	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_GetParticleLifetime(ulong)", GetParticleLifetime);
	MonoScriptingLayer::AddInternalCall("ParticleEmitterComponent::i_SetParticleLifetime(ulong,float)", SetParticleLifetime);
}

File: MonoScripting/Api/EngineClass/EntityComponents/ParticleEmitterComponent.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

public class ParticleEmitterComponent : IEntityComponent {
    public ParticleEmitterComponent(ulong handle) : base(handle)
    {
    }

    public ParticleEmitterComponent() : base(0)
    {
    }

    public void StartSimulation(){
        i_StartSimulation(handle);
    }
    public void StopSimulation(){
        i_StopSimulation(handle);
    }
    public bool IsSimulating(){
        return i_IsSimulating(handle);
    }

    public float GetDuration(){
        return i_GetDuration(handle);
    }
    public void SetDuration(float duration){
        i_SetDuration(handle, duration);
    }

    public bool IsLooping(){
        return i_IsLooping(handle);
    }
    public void SetLooping(bool looping){
        i_SetLooping(handle, looping);
    }

    public bool IsPrewarm(){
        return i_IsPrewarm(handle);
    }
    public void SetPrewarm(bool prewarm){
        i_SetPrewarm(handle, prewarm);
    }

    public float GetParticleLifetime(){
        return i_GetParticleLifetime(handle);
    }
    public void SetParticleLifetime(float lifetime){
        i_SetParticleLifetime(handle, lifetime);
    }

    public uint GetMaxParticlesCount(){
        return i_GetMaxParticlesCount(handle);
    }
    public void SetMaxParticlesCount(uint max_particles){
        i_SetMaxParticlesCount(handle, max_particles);
    }

    public void SetEmissionRate(int min, int max){

    }

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_StartSimulation(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_StopSimulation(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern bool i_IsSimulating(ulong handle);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern float i_GetDuration(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetDuration(ulong handle, float duration);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern bool i_IsLooping(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetLooping(ulong handle, bool looping);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern bool i_IsPrewarm(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetPrewarm(ulong handle, bool prewarm);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern float i_GetParticleLifetime(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetParticleLifetime(ulong handle, float lifetime);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern uint i_GetMaxParticlesCount(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetMaxParticlesCount(ulong handle, uint max_particles);
}